instrumentation/action_pack/lib/opentelemetry/instrumentation/action_pack/patches/action_controller/metal.rb

```diff
@@ -26,11 +31,10 @@ def dispatch(name, request, response)
 
             private
 
-            def add_rails_route(rack_span, request)
-              ::Rails.application.routes.router.recognize(request) do |route, _params|
-                rack_span.set_attribute('http.route', route.path.spec.to_s)
+            def rails_route(request)
+              @rails_route ||= ::Rails.application.routes.router.recognize(request) do |route, _params|
+                return route.path.spec.to_s
```

Comment:
For context I'm investigating a bug where I have evidence that this PR cause `request.script_name = "/"`, resulting in duplicate `/` in `request.url` e.g. `request.url => "http://example.com//foo"`.

I haven't fully pinpointed what is going on, but in the meantime this change seem very suspect.

If you `return` here then this memoization is never gonna work. So might be totally orthogonal, but this is very fishy.

Reply:
So what is happening in our app is that we reach this line: https://github.com/rails/rails/blob/a51e9676e611f578bdfd10fcf6891d9e2d762a23/actionpack/lib/action_dispatch/journey/router.rb#L68

We end up matching a catch all engine (`#<ActionDispatch::Journey::Route:0x000000012f5ae0d0 @name="security_reports_engine" ...`) which isn't anchored, and cause this bug.

Overall I don't think it's safe to call `recognize` where it's not expected.

Reply:
@casperisfine do you have a suggestion for what this should do instead or is it unsafe to use recognize_route at all?

Reply:
I'm not that familiar with the router, but yeah I'd suggest to wait for the router to be called by rails and grab the route there instead. (just a guess, might not be possible).

We just turned off that option.